src/material-experimental/form-field/testing/form-field-harness.ts

```diff
@@ -0,0 +1,229 @@
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {
+  ComponentHarness,
+  HarnessEnvironment,
+  HarnessPredicate,
+  TestElement
+} from '@angular/cdk/testing';
+import {MatInputHarness} from '@angular/material-experimental/input/testing';
+import {MatSelectHarness} from '../../mdc-select/harness/select-harness';
+import {FormFieldHarnessFilters} from './form-field-harness-filters';
+
+// TODO(devversion): support datepicker harness once developed (COMP-203).
+// Also support chip list harness.
+/** Possible harnesses of controls which can be bound to a form-field. */
+export type FormFieldControlHarness = MatInputHarness|MatSelectHarness;
+
+/**
+ * Harness for interacting with a standard Material form-field's in tests.
+ * @dynamic
+ */
+export class MatFormFieldHarness extends ComponentHarness {
+  static hostSelector = '.mat-form-field';
+
+  /**
+   * Gets a `HarnessPredicate` that can be used to search for an form-field with
+   * specific attributes.
+   * @param options Options for narrowing the search:
+   *   - `selector` finds a form-field that matches the given selector.
+   * @return a `HarnessPredicate` configured with the given options.
+   */
+  static with(options: FormFieldHarnessFilters = {}): HarnessPredicate<MatFormFieldHarness> {
+    return new HarnessPredicate(MatFormFieldHarness, options);
+  }
+
+  private _prefixContainer = this.locatorForOptional('.mat-form-field-prefix');
+  private _suffixContainer = this.locatorForOptional('.mat-form-field-suffix');
+  private _label = this.locatorForOptional('.mat-form-field-label');
+  private _errors = this.locatorForAll('.mat-error');
+  private _hints = this.locatorForAll('mat-hint,.mat-hint');
+
+  private _controlHarnessSelector =
+      [MatInputHarness, MatSelectHarness].map(h => h.hostSelector).join(',');
+
+  /** Gets the appearance of the form-field. */
+  async getAppearance(): Promise<'legacy'|'standard'|'fill'|'outline'> {
+    const hostEl = await this.host();
+    const [isLegacy, isStandard, isFill, isOutline] = await Promise.all([
+      hostEl.hasClass('mat-form-field-appearance-legacy'),
+      hostEl.hasClass('mat-form-field-appearance-standard'),
+      hostEl.hasClass('mat-form-field-appearance-fill'),
+      hostEl.hasClass('mat-form-field-appearance-outline'),
+    ]);
+    if (isLegacy) {
+      return 'legacy';
+    } else if (isStandard) {
+      return 'standard';
+    } else if (isFill) {
+      return 'fill';
+    } else if (isOutline) {
+      return 'outline';
+    }
+    throw Error('Could not determine appearance of form-field.');
+  }
+
+  /** Gets the harness of the control that is bound to the form-field. */
+  async getControl(): Promise<FormFieldControlHarness|null> {
+    const environment: HarnessEnvironment<any> = (this as any).locatorFactory;
+    // TODO(reviewer): not sure what we should do here. We need to construct a combined
```

Comment:
I need feedback on this. This is just a hack for now do demonstrate on what I have in mind. Any other ideas/or should we expose these things in the locator factory?

Reply:
Adding what I wrote on Slack. This would be my proposal:
```ts
private _controlSelector = this.locatorForAll('input,select');

// ----

const elements = await this._controlSelector();

if (!elements.length) {
  return null;
}
...
this.environment.createHarness(elements[0], HarnessType);
```

Basically something like that. It does not introduce too much changes in the harness infrastructure and the mental model remains quite similar IMO.

Two things are special here:
* Exposing the environment to the harnesses, so that they can create harnesses out of elements.
* And the other interesting thing is that createHarness creates the harness with an actual TestElement. It can do that because the corresponding environment knows how it can access the native element/ or it can just re-use the passed element.

Reply:
Wouldn't it be much simpler to have mat-form-field apply a marker CSS class to its control and look for that?

Reply:
Yeah that would be another possibility. Though it feels slightly wrong to add host bindings (which will be checked by Angular and affect micro-perf) in favor of a test harness.

I have the feeling that this is a pattern that people will hit more often (i.e. creating a harness of an element), so I'd vote for adding something like that anyway.

Reply:
It wouldn't be a binding, just a static class we add once here:
https://github.com/angular/components/blob/94c3fed9da74f3560467a4d702c240120a4c4c7f/src/material/form-field/form-field.ts#L300-L299

Reply:
Should we still keep track (somewhere else) of the use-case of creating a harness out of an element? cc. @mmalerba

Reply:
@devversion yeah why don't you file an issue to consider adding the feature and we'll see if anyone chimes in asking for it. It feels like a pretty niche case to me, so we may be able to do without it.

Reply:
Sounds good. will do.